fix: RouterComponent replace methods to correctly handle previous/nextRoute

spydon opened the pull request:
<!--
The title of your PR on the line above should start with a [Conventional Commit] prefix
(`fix:`, `feat:`, `docs:`, `test:`, `chore:`, `refactor:`, `perf:`, `build:`, `ci:`,
`style:`, `revert:`). This title will later become an entry in the [CHANGELOG], so please
make sure that it summarizes the PR adequately.
-->

# Description

Before when you only had two routes you'd get:
```
Bad state: No element

dart:core                                                  List.last
package:flame/src/components/router_component.dart 260:30  RouterComponent._popWithoutMinimumAssert
package:flame/src/components/router_component.dart 122:7   RouterComponent.pushReplacementNamed
test/components/router_component_test.dart 20:14           main.<fn>.<fn>
```


## Checklist
<!--
Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes with `[x]`. If some checkbox is not applicable, mark it as `[-]`.
-->

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.


## Breaking Change?
<!--
Would your PR require Flame users to update their apps following your change?

If yes, then the title of the PR should include "!" (for example, `feat!:`, `fix!:`). See
[Conventional Commit] for details. Also, for a breaking PR uncomment and fill in the "Migration
instructions" section below.

### Migration instructions

If the PR is breaking, uncomment this header and add instructions for how to migrate from the
currently released version to the new proposed way.
-->

- [x] No, this PR is not a breaking change.


## Related Issues
<!--
Indicate which issues this PR resolves, if any. For example:

Closes #1234
!-->

<!-- Links -->
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

munsterlander reviewed (commented):
LGTM - odd that made it through, but it comes from pop with the same tests just modified.

Edit to clarify:

I just want to confirm that `pop()` does not have the same issue. 

```dart
void pop() {
    assert(
      _routeStack.length > 1,
      'Cannot pop the last route from the Router',
    );
    final route = _routeStack.removeLast();
    _adjustRoutesOrder();
    _adjustRoutesVisibility();
    route.didPop(_routeStack.last);
}
```

spydon commented:
> I just want to confirm that `pop()` does not have the same issue.

It does indeed, good catch!
I'll write some tests that tests this.

spydon commented:
Ah wait, this is wrong, `onPop` should return the next route, not the route that was popped.

munsterlander commented:
> Ah wait, this is wrong, `onPop` should return the next route, not the route that was popped.

Ah! When I was adding this, I was confused why that was there, but not knowing the full router creation origin story, I was like well there had to be a reason.   The tests passed for what wa expected, but good to know.

spydon commented:
> Ah! When I was adding this, I was confused why that was there, but not knowing the full router creation origin story, I was like well there had to be a reason. The tests passed for what was expected, but good to know.

Fixed it and added some extra tests that should cover this. :)

st-pasha commented on packages/flame/lib/src/components/router_component.dart:
```diff
@@ -255,9 +239,9 @@ class RouterComponent extends Component {
 
   /// Local method to bypass [pop]'s assert
-  void _popWithoutMinimumAssert() {
+  void _replaceRoute(Route nextRoute) {
```
`_removeTopRoute`? This method doesn't actually replace anything.

spydon replied:
Done

st-pasha commented on packages/flame/test/components/router_component_test.dart:
```diff
@@ -4,6 +4,47 @@ import 'package:flame_test/flame_test.dart';
 import 'package:flutter/widgets.dart' hide Route, OverlayRoute;
 import 'package:flutter_test/flutter_test.dart';
 
+class _TestRoute extends Route {
```
Could you please move this helper class to the bottom of the file? That way, when you open file you can immediately see `void main()` and can press the green arrow to start the tests.

spydon replied:
Done
